### MINI-Project-3

#### Where is the data
https://data.cityofnewyork.us/Housing-Development/apartment-cost-list/wnmh-ssrd/about_data

In [75]:
import pandas as pd
import numpy as np
from runEDAFunctions import *
from RunAllRegressionMLs import *
from sklearn.preprocessing import LabelEncoder, StandardScaler
import importlib
import runEDAFunctions 
import RunAllRegressionMLs

importlib.reload(runEDAFunctions)
importlib.reload(RunAllRegressionMLs)


<module 'RunAllRegressionMLs' from '/Users/pnandan/Documents/Data-Science-Project/Mini-Project-3/RunAllRegressionMLs.py'>

In [76]:
## Read the CSV
ny_df = pd.read_csv('apartment_cost_list_20240814.csv')

In [77]:
ny_df.shape

(19601, 18)

In [78]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19601 entries, 0 to 19600
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Job #                   19601 non-null  int64 
 1   Borough                 19601 non-null  object
 2   House #                 19601 non-null  object
 3   Street Name             19601 non-null  object
 4   Block                   19601 non-null  int64 
 5   Lot                     19601 non-null  int64 
 6   Bin #                   19601 non-null  int64 
 7   Job Type                19601 non-null  object
 8   Community - Board       19601 non-null  int64 
 9   Curb Cut                16 non-null     object
 10  Fully Permitted         15633 non-null  object
 11  Initial Cost            19601 non-null  object
 12  Proposed Zoning Sqft    19601 non-null  int64 
 13  Horizontal Enlrgmt      163 non-null    object
 14  Vertical Enlrgmt        242 non-null    object
 15  En

In [79]:
ny_df.head(10)

Job #    Borough House #        Street Name  Block   Lot    Bin #  \
0  140922893  MANHATTAN     160            E 89 ST   1517    43  1048094   
1  301773579   BROOKLYN      65       GATES AVENUE   1962    86  3056039   
2  302000974   BROOKLYN     501       HICKS STREET    321  7501  3003731   
3  302002990   BROOKLYN     334     MILFORD STREET   4056   145  3393496   
4  302008645   BROOKLYN     611     HANCOCK STREET   1656    59  3045842   
5  141002313  MANHATTAN     242     EAST 19 STREET    899    32  1019591   
6  302022656   BROOKLYN       6           2 STREET    460     2  3007811   
7  102636578  MANHATTAN     192      SPRING STREET    489    20  1007355   
8  302026849   BROOKLYN    1054  WILLOUGHBY AVENUE   3208    11  3325441   
9  102652791  MANHATTAN     194    RIVERSIDE DRIVE   1251    71  1034108   

  Job Type  Community - Board Curb Cut Fully Permitted Initial Cost  \
0       A2                108      NaN      07/14/2020    $85000.00   
1       A1                302      NaN      05/23/2007    $60000.00   
2       A2                306      NaN      12/12/2005    $40000.00   
3       A1                305      NaN      01/06/2006    $65000.00   
4       A1                303      NaN      11/28/2005    $50000.00   
5       A2                106      NaN      06/23/2021    $44000.00   
6       A2                306      NaN      04/23/2008    $60000.00   
7       A2                102      NaN      03/01/2001    $50000.00   
8       A2                304      NaN             NaN    $40000.00   
9       A2                107      NaN             NaN    $50000.00   

   Proposed Zoning Sqft Horizontal Enlrgmt Vertical Enlrgmt  \
0                     0                NaN              NaN   
1                     0                NaN              NaN   
2                     0                NaN              NaN   
3                     0                NaN              NaN   
4                     0                NaN              NaN   
5                     0                NaN              NaN   
6                     0                NaN              NaN   
7                     0                NaN              NaN   
8                     0                NaN              NaN   
9                     0                NaN              NaN   

   Enlargement SQ Footage Zoning Dist1  \
0                       0          R8B   
1                       0           R6   
2                       0           R6   
3                       0           R6   
4                       0           R5   
5                       0          R9A   
6                       0          NaN   
7                       0         C1-5   
8                       0           R6   
9                       0         R10A   

                                     Job Description  
0  HEREBY FILING APPLICATION FOR THE INTERIOR REN...  
1  Change existing four story and basement brick ...  
2  ERECTION OF MEZZANINE IN EXISTING APARTMENT #4...  
3  Proposed convert 1 Duplex  apartment (UG.J3) i...  
4  APPLICATION FILED TO CONVERT FROM 2 FAMILY TO ...  
5  STRUCTURAL WORK FOR NEW CONVENIENCE STAIR IN C...  
6  RENOVATE EXISTING LOWER DUPLEX APARTMENT. NO C...  
7  COMBINE APARTMENTS AT 5TH FLOOR, APARTMENTS 19...  
8  Application being filed to renovate existing s...  
9  Interior renovation to Apartment 2F. Close ope...

In [80]:
# Calling functions from runEDAFunctions.py file to check which columns has nulls
print(get_columns_with_nulls(ny_df))

[{'Curb Cut': 19585}, {'Fully Permitted': 3968}, {'Horizontal Enlrgmt': 19438}, {'Vertical Enlrgmt': 19359}, {'Zoning Dist1': 2096}]


In [81]:
ny_df['Fully Permitted'].unique()

array(['07/14/2020', '05/23/2007', '12/12/2005', ..., '12/21/2006',
       '05/28/2004', '09/29/2004'], dtype=object)

In [82]:
random_date_choices = np.array(ny_df['Fully Permitted'].unique())
random_date_choices = random_date_choices[~pd.isna(random_date_choices)] ## Filter Null value from the unique list

In [83]:
ny_df = replace_nulls_with_random(ny_df, 'Fully Permitted', random_date_choices)

In [84]:
ny_df.isnull().sum()

Job #                         0
Borough                       0
House #                       0
Street Name                   0
Block                         0
Lot                           0
Bin #                         0
Job Type                      0
Community - Board             0
Curb Cut                  19585
Fully Permitted               0
Initial Cost                  0
Proposed Zoning Sqft          0
Horizontal Enlrgmt        19438
Vertical Enlrgmt          19359
Enlargement SQ Footage        0
Zoning Dist1               2096
Job Description               0
dtype: int64

In [85]:
ny_df['Zoning Dist1'].unique()

array(['R8B', 'R6', 'R5', 'R9A', nan, 'C1-5', 'R10A', 'C4-6A', 'C5-1',
       'R10', 'R10-', 'M1-6', 'C1-8X', 'R8A', 'C4-5', 'R6A', 'R6B', 'R8',
       'M1-1', 'C4-3', 'R7A', 'R7B', 'C1-6', 'C1-1', 'R4', 'C1-3', 'C6-4',
       'C1-9A', 'R7-1', 'C2-6', 'R8X', 'C6-2A', 'C2-3', 'C1-2', 'C6-2',
       'C6-2G', 'R5B', 'R3-2', 'R7-2', 'C1-6A', 'C6-1', 'M1-4', 'C6-2M',
       'M1-5M', 'M1-2/R6B', 'C4-4A', 'M1-5B', 'R10H', 'C1-9', 'M1-5',
       'M1-2', 'R4A', 'R3-1', 'C1-4', 'C5-5', 'C2-4', 'C4-2', 'M3-1',
       'C4-2A', 'R4-1', 'C2-2', 'M1-2/R6A', 'R9', 'R3A', 'C2-8', 'C1-7A',
       'C4-3A', 'C6-6', 'C2-8A', 'C6-4M', 'C6-3A', 'C6-4A', 'C5-2',
       'C2-7A', 'C5-3', 'M1-5A', 'M1-3D', 'C6-3', 'C5-P', 'R1-2', 'C8-4',
       'R9X', 'C6-6.5', 'C6-4X', 'C1-7', 'C5-1A', 'C4-7', 'C6-3X', 'M2-1',
       'BPC', 'C1-8A', 'C4-4', 'C6-1G', 'C6-9', 'C5-2.5', 'C5-2A', 'C6-5',
       'C6-7T', 'C8-2', 'M1-2/R6', 'R3X', 'C4-6', 'M1-2/R5D', 'M1-4/R6B',
       'M1-6D', 'C4-4D', 'M1-4D', 'M1-2/R8A', 'R5D', 'C

In [86]:
random_zone_dist = np.array(ny_df['Zoning Dist1'].unique())
random_zone_dist = random_zone_dist[~pd.isna(random_zone_dist)] ## Filter Null value from the unique list

In [87]:
ny_df = replace_nulls_with_random(ny_df, 'Zoning Dist1', random_zone_dist)

In [88]:
ny_df.isnull().sum()

Job #                         0
Borough                       0
House #                       0
Street Name                   0
Block                         0
Lot                           0
Bin #                         0
Job Type                      0
Community - Board             0
Curb Cut                  19585
Fully Permitted               0
Initial Cost                  0
Proposed Zoning Sqft          0
Horizontal Enlrgmt        19438
Vertical Enlrgmt          19359
Enlargement SQ Footage        0
Zoning Dist1                  0
Job Description               0
dtype: int64

#### There are too many nulls for {'Curb Cut': 19585},{'Horizontal Enlrgmt': 19438}, {'Vertical Enlrgmt': 19359} out of total "19601"
hence as a treatment, we will drop these columns

In [89]:
ny_df.drop(['Curb Cut','Horizontal Enlrgmt','Vertical Enlrgmt'], axis=1, inplace=True)

In [90]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19601 entries, 0 to 19600
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Job #                   19601 non-null  int64 
 1   Borough                 19601 non-null  object
 2   House #                 19601 non-null  object
 3   Street Name             19601 non-null  object
 4   Block                   19601 non-null  int64 
 5   Lot                     19601 non-null  int64 
 6   Bin #                   19601 non-null  int64 
 7   Job Type                19601 non-null  object
 8   Community - Board       19601 non-null  int64 
 9   Fully Permitted         19601 non-null  object
 10  Initial Cost            19601 non-null  object
 11  Proposed Zoning Sqft    19601 non-null  int64 
 12  Enlargement SQ Footage  19601 non-null  int64 
 13  Zoning Dist1            19601 non-null  object
 14  Job Description         19601 non-null  object
dtypes:

In [91]:
## Find duplicates
ny_df.duplicated().sum()

4631

#### Lets drop duplicates

In [92]:
ny_df = ny_df.drop_duplicates()

In [93]:
ny_df.shape

(14970, 15)

In [94]:
ny_df.describe().T

count          mean           std          min  \
Job #                   14970.0  1.616443e+08  9.138209e+07  102133077.0   
Block                   14970.0  1.556229e+03  1.552652e+03          0.0   
Lot                     14970.0  1.275082e+03  2.776938e+03          0.0   
Bin #                   14970.0  1.500749e+06  9.282070e+05    1000043.0   
Community - Board       14970.0  1.507103e+02  8.971049e+01        101.0   
Proposed Zoning Sqft    14970.0  7.724707e+02  1.160819e+04          0.0   
Enlargement SQ Footage  14970.0  1.969299e+01  3.361601e+02          0.0   

                                25%          50%          75%          max  
Job #                   104710617.0  121694880.5  140790745.0  540180276.0  
Block                         821.0       1248.0       1566.0      16278.0  
Lot                            17.0         36.0         67.0       9007.0  
Bin #                     1028820.0    1046957.0    1084086.5    5860016.0  
Community - Board             105.0        107.0        110.0        503.0  
Proposed Zoning Sqft            0.0          0.0          0.0     724475.0  
Enlargement SQ Footage          0.0          0.0          0.0      29687.0

In [95]:
ny_df['Initial Cost'].describe()

count         14970
unique           46
top       $40000.00
freq           2194
Name: Initial Cost, dtype: object

#### Convert Job type to Numerical using Label Encoding

In [96]:
le = LabelEncoder()
ny_df['Job Type'] = le.fit_transform(ny_df['Job Type'])

In [97]:
ny_df['Initial Cost'] = ny_df['Initial Cost'].apply(lambda x: float(x.replace('$','').replace(',','')))

In [98]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14970 entries, 0 to 19600
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Job #                   14970 non-null  int64  
 1   Borough                 14970 non-null  object 
 2   House #                 14970 non-null  object 
 3   Street Name             14970 non-null  object 
 4   Block                   14970 non-null  int64  
 5   Lot                     14970 non-null  int64  
 6   Bin #                   14970 non-null  int64  
 7   Job Type                14970 non-null  int64  
 8   Community - Board       14970 non-null  int64  
 9   Fully Permitted         14970 non-null  object 
 10  Initial Cost            14970 non-null  float64
 11  Proposed Zoning Sqft    14970 non-null  int64  
 12  Enlargement SQ Footage  14970 non-null  int64  
 13  Zoning Dist1            14970 non-null  object 
 14  Job Description         14970 non-null  obj

In [99]:
print(ny_df['Initial Cost'].unique())
print(ny_df['Job Type'].unique())

[ 85000.  60000.  40000.  65000.  50000.  44000.  69000.  55000.  52000.
  45000.  75000.  62000.  71000.  42000.  70000.  61000.  80000.  41000.
  82000.  67000.  72000.  58000.  64000.  66000.  53000.  77000.  73000.
 480000.  54000.  63000.  57000.  79000.  68000.  56000.  43000.  84000.
  51000.  59000.  81000.  83000.  76000.  78000. 470000.  74000. 490000.
 460000.]
[1 0 2]


#### Separate Year From Date

In [100]:
from datetime import datetime
ny_df = pd.DataFrame(ny_df)
ny_df['date']=pd.to_datetime(ny_df['Fully Permitted'])
ny_df['year']=ny_df['date'].dt.year


In [101]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14970 entries, 0 to 19600
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Job #                   14970 non-null  int64         
 1   Borough                 14970 non-null  object        
 2   House #                 14970 non-null  object        
 3   Street Name             14970 non-null  object        
 4   Block                   14970 non-null  int64         
 5   Lot                     14970 non-null  int64         
 6   Bin #                   14970 non-null  int64         
 7   Job Type                14970 non-null  int64         
 8   Community - Board       14970 non-null  int64         
 9   Fully Permitted         14970 non-null  object        
 10  Initial Cost            14970 non-null  float64       
 11  Proposed Zoning Sqft    14970 non-null  int64         
 12  Enlargement SQ Footage  14970 non-null  int64      

In [102]:
ny_df.head()

Job #    Borough House #     Street Name  Block   Lot    Bin #  \
0  140922893  MANHATTAN     160         E 89 ST   1517    43  1048094   
1  301773579   BROOKLYN      65    GATES AVENUE   1962    86  3056039   
2  302000974   BROOKLYN     501    HICKS STREET    321  7501  3003731   
3  302002990   BROOKLYN     334  MILFORD STREET   4056   145  3393496   
4  302008645   BROOKLYN     611  HANCOCK STREET   1656    59  3045842   

   Job Type  Community - Board Fully Permitted  Initial Cost  \
0         1                108      07/14/2020       85000.0   
1         0                302      05/23/2007       60000.0   
2         1                306      12/12/2005       40000.0   
3         0                305      01/06/2006       65000.0   
4         0                303      11/28/2005       50000.0   

   Proposed Zoning Sqft  Enlargement SQ Footage Zoning Dist1  \
0                     0                       0          R8B   
1                     0                       0           R6   
2                     0                       0           R6   
3                     0                       0           R6   
4                     0                       0           R5   

                                     Job Description       date  year  
0  HEREBY FILING APPLICATION FOR THE INTERIOR REN... 2020-07-14  2020  
1  Change existing four story and basement brick ... 2007-05-23  2007  
2  ERECTION OF MEZZANINE IN EXISTING APARTMENT #4... 2005-12-12  2005  
3  Proposed convert 1 Duplex  apartment (UG.J3) i... 2006-01-06  2006  
4  APPLICATION FILED TO CONVERT FROM 2 FAMILY TO ... 2005-11-28  2005

In [103]:
ny_df.shape

(14970, 17)

#### Plotting to understand the data

In [104]:
import plotly.express as px
fig = px.scatter(ny_df, x='Initial Cost', y = 'Proposed Zoning Sqft')
fig.show()

In [105]:
import plotly.express as px
fig = px.scatter(ny_df, x='Initial Cost', y = 'year')
fig.show()

In [106]:
import plotly.express as px
fig = px.histogram(ny_df, x='Job Type')
fig.show()

In [107]:
import plotly.express as px
fig = px.histogram(ny_df, x='Initial Cost')
fig.show()

In [108]:
ny_df = ny_df.rename(columns={'Job #': 'Job'})

In [109]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14970 entries, 0 to 19600
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Job                     14970 non-null  int64         
 1   Borough                 14970 non-null  object        
 2   House #                 14970 non-null  object        
 3   Street Name             14970 non-null  object        
 4   Block                   14970 non-null  int64         
 5   Lot                     14970 non-null  int64         
 6   Bin #                   14970 non-null  int64         
 7   Job Type                14970 non-null  int64         
 8   Community - Board       14970 non-null  int64         
 9   Fully Permitted         14970 non-null  object        
 10  Initial Cost            14970 non-null  float64       
 11  Proposed Zoning Sqft    14970 non-null  int64         
 12  Enlargement SQ Footage  14970 non-null  int64      

### Run machine learning algorithms

In [110]:

target_var = 'Initial Cost'
#feature_list = [col for col in ny_df.columns if col != 'Initial Cost']
feature_list = ['Job','Lot','Job Type','Proposed Zoning Sqft','Enlargement SQ Footage','year']
models = ['LinearRegression', 'Ridge', 'Lasso', 'DecisionTree', 'RandomForest', 'KNeighbors', 'SVR', 'AdaBoost', 'XGBoost']
# Run  for all models
evaluate_regression_models(models, ny_df, target_var, feature_list)

model type == LinearRegression  ..STARRING - evaluate_regression_model func
Model: LinearRegression
The RMSE on train data is  25811.772842327766
The RMSE on test data is  27025.125607769645
The MSE on train data is  666247617.2639292
The MSE on test data is  730357414.1157267
The MAE on train data is  12934.179401770873
The MAE on test data is  13201.722024982837
The R2 on train data is  0.0020933243643431743
The R2 on test data is  0.0005434030559255953
The Adjusted R2 on train data is  0.0015215672927414436
The Adjusted R2 on test data is  -0.0007939608115286045
Execution time: 0.04731607437133789 seconds
model type == LinearRegression  ..END - valuate_regression_model func
model type == Ridge  ..STARRING - evaluate_regression_model func
Model: Ridge
The RMSE on train data is  25811.77621991123
The RMSE on test data is  27024.904059624358
The MSE on train data is  666247791.6267749
The MSE on test data is  730345439.4319011
The MAE on train data is  12933.914391961924
The MAE on tes